File: csharp/src/Int2Dds/Conditions/GuardCondition.cs
using System;
using Int2Dds.Exceptions;
using Int2Dds.Interop;

namespace Int2Dds.Conditions
{
    /// <summary>
    /// A user-controlled condition for signaling.
    /// GuardConditions can be attached to a WaitSet and manually triggered
    /// to wake up waiting threads.
    /// </summary>
    public sealed class GuardCondition : IDisposable
    {
        private IntPtr _handle;
        private bool _disposed;

        /// <summary>
        /// Creates a new GuardCondition.
        /// </summary>
        public GuardCondition()
        {
            ReturnCodeHelper.CheckReturn(NativeMethods.int2dds_guard_condition_new(out _handle));
        }

        /// <summary>
        /// Gets the native handle for this guard condition.
        /// </summary>
        internal IntPtr Handle
        {
            get
            {
                if (_disposed) throw new ObjectDisposedException(GetType().Name);
                return _handle;
            }
        }

        /// <summary>
        /// Gets or sets the trigger value.
        /// </summary>
        public bool TriggerValue
        {
            get
            {
                if (_disposed) throw new ObjectDisposedException(GetType().Name);
                ReturnCodeHelper.CheckReturn(
                    NativeMethods.int2dds_guard_condition_get_trigger_value(_handle, out bool value));
                return value;
            }
            set
            {
                if (_disposed) throw new ObjectDisposedException(GetType().Name);
                ReturnCodeHelper.CheckReturn(
                    NativeMethods.int2dds_guard_condition_set_trigger_value(_handle, value));
            }
        }

        /// <summary>
        /// Sets the trigger value to true.
        /// </summary>
        public void Trigger() => TriggerValue = true;

        /// <summary>
        /// Sets the trigger value to false.
        /// </summary>
        public void Reset() => TriggerValue = false;

        public void Dispose()
        {
            if (_disposed) return;
            _disposed = true;

            if (_handle != IntPtr.Zero)
            {
                NativeMethods.int2dds_guard_condition_delete(_handle);
                _handle = IntPtr.Zero;
            }
        }
    }
}

File: java/api/src/main/java/com/intellectus/int2dds/conditions/GuardCondition.java
package com.intellectus.int2dds.conditions;

import com.intellectus.int2dds.internal.NativeKeepAlive;
import com.intellectus.int2dds.internal.ReturnCodes;
import com.intellectus.int2dds.internal.ffi.FfiAccess;

/** An application-triggered condition. */
public final class GuardCondition extends Condition {
    public GuardCondition() {
        super(create(), FfiAccess::guardConditionDelete);
    }

    private static long create() {
        long[] out = new long[1];
        int rc = FfiAccess.guardConditionNew(out);
        ReturnCodes.check(rc);
        return out[0];
    }

    /** Sets the trigger value; a WaitSet waiting on this wakes when true. */
    public void setTriggerValue(boolean value) {
        long h = handle();
        FfiAccess.guardConditionSetTrigger(h, value);
        NativeKeepAlive.keepAlive(this);
    }

    @Override
    public boolean triggerValue() {
        long h = handle();
        boolean[] out = new boolean[1];
        int rc = FfiAccess.guardConditionGetTriggerValue(h, out);
        NativeKeepAlive.keepAlive(this);
        ReturnCodes.check(rc);
        return out[0];
    }
}
